### PR TITLE
New speedreader tune dialog

### PR DESCRIPTION
**Pull request by boocmp**
Resolves https://github.com/brave/brave-browser/issues/31583

**1. The customization button in the toolbar:**
image

**2. Dialog box in the location bar when right-clicking:**
image

**3. Left-clicking toggles the speedreader without a dialog**

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: app/brave_generated_resources.grd [hunk 1]
```diff
@@ -788,43 +788,6 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
         </message>
       </if>
 
-      <!-- Speedreader -->
```
Comment (boocmp) [inline app/brave_generated_resources.grd:791]: Some is deleted, some is moved to speedreader_strings.grdp

File: components/speedreader/speedreader_util.cc [hunk 2]
```diff
@@ -6,66 +6,132 @@
...
+    return TransitToOriginalDirection(state, std::move(desired));
+  } else if (absl::holds_alternative<DistillStates::Pending>(desired)) {
+    return TransitToDistilledDirection(
+        state, absl::get<DistillStates::Pending>(desired).reason);
+  } else {
+    CHECK(absl::holds_alternative<DistillStates::Distilled>(desired));
+    return TransitToDistilledDirection(
+        state, DistillStates::Pending::Reason::kNone,
+        absl::get<DistillStates::Distilled>(desired).result);
+  }
```
[the remark this answers is not in the document]
Comment (boocmp) [reply components/speedreader/speedreader_util.cc:118]: I looked at the typos.yaml and didn't understand why it triggered here.

[the remark this answers is not in the document]
Comment (fmarier) [reply components/speedreader/speedreader_util.cc:118]: I also don't see why it triggered.

File: components/resources/speedreader_strings.grdp [hunk 3]
```diff
@@ -1,31 +1,45 @@
...
-    <message name="IDS_READER_MODE_SPEEDREADER" desc="Tooltip for disable/enable speedreader button">Speedreader mode toggle</message>
-    <message name="IDS_READER_MODE_AI" desc="Tooltop for AI button">Brave AI</message>
-  </if>
+    <message name="IDS_SPEEDREADER_SHOW_ORIGINAL_PAGE_LINK" desc="Text of the link to show original page">View original</message>
+    <message name="IDS_READER_MODE_MINUTES_TEXT" desc="Text of the read time element">min. read</message>
+
+    <if expr="not is_android">
+      <message name="IDS_SPEEDREADER_ICON_TURN_OFF_READER_MODE" desc="Accessible text shown when a page is in Reader Mode">Turn off speedreader</message>
+      <message name="IDS_SPEEDREADER_ICON_TURN_ON_READER_MODE" desc="Accessible text shown when a page can be distilled">Turn on speedreader</message>
+      <message name="IDS_READER_MODE_SHOW_ORIGINAL_PAGE_LINK" desc="Text of the link to show original page">View original</message>
```
Comment (mkarolin) [inline components/resources/speedreader_strings.grdp:10]: Is this used? Is it meant to replace `IDS_SPEEDREADER_SHOW_ORIGINAL_PAGE_LINK` (above, line 4)?

Comment (boocmp) [reply components/resources/speedreader_strings.grdp]: No, it doesn't. removed.

**Review by mkarolin — APPROVED** [review]
`strings`++

File: browser/ui/views/speedreader/reader_mode_bubble.cc [hunk 4]
```diff
@@ -108,53 +85,107 @@ void ReaderModeBubble::Init() {
...
+      box->SetBorder(
+          views::CreateThemedSolidSidedBorder(border, ui::kColorMenuSeparator));
+    }
+
+    auto label = std::make_unique<views::Label>();
+    label->SetText(brave_l10n::GetLocalizedResourceUTF16String(ids));
+    label->SetFontList(font);
+    label->SetLineHeight(kToggleLineHeight);
+    label->SetMultiLine(true);
+    // label->SetMaximumWidth(kToggleWidth);
```
Comment (nullhook) [inline browser/ui/views/speedreader/reader_mode_bubble.cc:114]: leftover?

**Review by nullhook — CHANGES_REQUESTED** [review]

File: browser/ui/views/frame/brave_browser_view.cc [hunk 5]
```diff
@@ -385,23 +380,26 @@ void BraveBrowserView::SetStarredState(bool is_starred) {
...
+      arrow = views::BubbleBorder::TOP_RIGHT;
+      break;
+    case speedreader::SpeedreaderBubbleLocation::kToolbar:
+      anchor = reader_mode_toolbar_view_->toolbar();
+      arrow = views::BubbleBorder::TOP_LEFT;
+      break;
+  }
+
+  auto* reader_mode_bubble =
+      new speedreader::ReaderModeBubble(anchor, tab_helper);
```
Comment (nullhook) [inline browser/ui/views/frame/brave_browser_view.cc:398]: I'm aware that this bubble wasn't introduced in this PR and maybe my feedback for this name would be out of scope. 

Shouldn't this class be "ReaderModeBubbleView"?

File: components/speedreader/resources/panel/components/lists/style.ts [hunk 6]
```diff
@@ -67,6 +67,7 @@ export const Button = styled.button<{inGroup?: boolean}>`
   background-color: transparent;
 
   &.is-active {
+    color: rgb(77, 82, 253);
```
[the remark this answers is not in the document]
Comment (boocmp) [reply components/speedreader/resources/panel/components/lists/style.ts:70]: There is no variable with this value, can I add it?  And how ?

File: components/speedreader/resources/panel/components/toolbar/index.tsx [hunk 7]
```diff
@@ -90,17 +86,22 @@ function Toolbar() {
     getToolbarAPI().dataHandler.aiChat()
   }
 
+  const handleTune = (show: boolean) => {
```
[the remark this answers is not in the document]
Comment (boocmp) [reply components/speedreader/resources/panel/components/toolbar/index.tsx:89]: The button is called "Tune", so handleTune means that it handles pressing the Tune button.

**Comment by iefremov** [comment]
@rebron we need to design new P3A metrics for speedreader, b/c the current ones are outdated. I'd suggest something like "How many times reader mode has been toggled during the last week", for the beginning. And, say, whether the global speedreader toggle has been used.

**Review by iefremov — APPROVED** [review]

File: components/speedreader/resources/panel/components/lists/style.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nullhook) [reply components/speedreader/resources/panel/components/lists/style.ts:71]: looks like the variable isn't available in the current version of leo yet. It's suppose to be: 
`--leo-color-button-background`

**Review by nullhook — APPROVED** [review]

**Review by wchen342 — APPROVED** [review]
Android++

**Comment by MadhaviSeelam** [comment]
Verification `In progress` using
```
Brave | 1.59.62 Chromium: 116.0.5845.163 (Official Build) nightly (64-bit)
-- | --
Revision | 0018a11e701393659831fe88dea09e0d53b21840
OS | Windows 11 Version 22H2 (Build 22621.2134)
```
#### Case 1: Speedreader button with new tune speedreader dialog in the URL bar  
<details>

1. Install 1.59.62
2. launch Brave
3. opened `brave://settings/appearance`
4. selected `Dark` in `Brave colors
5. verified `Speedreader` setting is `Off` 
6. navigated to nytimes.com and opened an article - https://www.nytimes.com/interactive/2023/08/31/travel/things-to-do-amsterdam.html
7. verified`Speedreader` button is available in the URL bar
8. hovered over and correct text displayed `Turn on speedreader` - File as a follow up- correct text `Speedreader`
9. clicked `Speedreader` button 
10. confirmed when Speedreader button is clicked, no other bubbles shown in the URL bar
11. confirmed article layout is shown in `Speedreader` 
12. right clicked `Speedreader` button in the URL bar
13. confirmed new `Tune speedreader` dialog is shown with default options `Off`
14. clicked/toggled `Off` speedreader button in the URL bar
15. confirmed the speedreader mode off and article is no longer shown in speedreader mode


step 4 | step 5 | step 7-8 | step 13 | step 15
-----|-----|-----|------|------
image|image|image|image|image
</details>

Case 2:  Speedreader buttons UI  
<details>

1. continue from Case 1
     - Speedreader turned on for an article in a new tab - https://www.nytimes.com/interactive/2023/08/31/travel/things-to-do-amsterdam.html
2. confirmed following `Speedreader` buttons/options are shown above the article and match to designs/screenshot in the issue and are in the correct order left to right
      - `Tune Speedreader` button
      - `Appearance settings` (Aa control)
      - `Brave Ai` 
      - `X` contro
2. confirmed the `Turn speedreader` icon matches the figma designs
3. clicked on `Tune Speedreader` button 
4. confirmed the default options are `off` as below in the `speedreader tune dialog`
       - `Always use speedreader for this site`
       - `Always use speedreader for all sites` 
5. confirmed `Tune speedreader dialog` is shown and matches to Figma designs/screenshots 

step 1 | step 5
------ | -----
image|image|image
</details>

Case 3: Tune speedreader dialog functionality
<details>

***_Default settings for `Always use speedreader for this site`  - `Off` _***

1. continue from Case 2
- `Turn speedreader` dialog is shown with default options on the speedreader buttons toolbar for the article https://www.nytimes.com/interactive/2023/08/31/travel/things-to-do-amsterdam.html
2. navigate to another article from the same site `nytimes.com` in a new tab https://www.nytimes.com/interactive/2023/03/09/travel/things-to-do-nashville.html
3. confirmed page is not shown in `Speedreader` mode

step 1 | step 2-3 
------ | ------
image|image|

***_Always use speedreader for this site` - enabled_***

1. continue from above `Default settings` 
2. enable `Always use speedreader for this site` for the first article - https://www.nytimes.com/interactive/2023/08/31/travel/things-to-do-amsterdam.html
3. return to second article and reload the article https://www.nytimes.com/interactive/2023/03/09/travel/things-to-do-nashville.html
4. confirmed the `Always use speedreader for this site` is carried to second article
5. load another article from the same site in a new tab - https://www.nytimes.com/interactive/2023/02/23/travel/things-to-do-miami.html
6. confirmed `Always use speedreader for this site` is carried to third article 
7. visit a different site sfchronicle.com and open a new article - https://www.sfchronicle.com/projects/best-ice-cream-sf/
8. confirmed `Always use speedreader for this site` setting is not carried to `sfchronicle.com` site. 

 step 2 | step 4 | step 6 | step 8
---- | ----- | ---- | -----  
image|image|image|image

</details>

**Comment by kjozwiak** [comment]
Changing this back to `1.59.x` as this landed in `master` and not `1.58.x`.

**Pull request merged**